### DISCUSSION TITLE
BL0939 becomes unuseable when third serial port on ESP32 is configured

### DISCUSSION THREAD

**Discussion by thomas374b** [arendst/Tasmota#20060] [Ideas]
### PROBLEM DESCRIPTION
The ESP32 has 3 hardware serial ports. However only one can be used simultaneously together with energy sensor BL0939

### REQUESTED INFORMATION
- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [ ] Device used (e.g., Sonoff Basic): ESP32 with BL0939 
- [ ] Tasmota binary firmware version number used: 
  - [ ] Pre-compiled
  - [x] Self-compiled: 13.2.0
- [ ] Flashing tools used: _____
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
- compile tasmota with defines   USE_MODBUS_BRIDGE_TCP, USE_TCP_BRIDGE, USE_ENERGY_SENSOR, USE_MODBUS_ENERGY,  USE_BL09XX  on main branch
- use an ESP32, connect a  BL0939 energy sensor and two serial interfaces to it
- configure the  "BL0939 Rx", "Serial Tx", "ZC pulse", "TCP Rx", "TCP Tx", "Modbus Tx", "Modbus Rx" and "Modbus Tx Enable" pins where you have connected the corresponding devices/interfaces  (Modbus pins 1,3,5) (Serial TCP Bridge 16,17) BL0939 (26,25,4)
- connect a power sink (e.g. light bulb)
- configure TCPBaudrate 2400 and TCPStart 987 on the console
- configure seriallog 0, weblog 4 on the console
- open console window in browser
- connect with "telnet tasmotahost 987" from PC to the ESP32
- connect with "modbus" tool from pypi to the ESP32

### EXPECTED BEHAVIOUR
- the energy sensor shows correct values for voltage, current and power
- on the console there are lines with "BL9: RX???????" logged (this should use the first serial port)
- data on the second serial port are forwarded from and to the network and show up in the telnet window
- the modbus bridge forwards packets from TCP to third serial port and vice versa

### ACTUAL BEHAVIOUR
- proper sensor data from BL0939 is only shown if only one another serial port is configured, either Modbus Rx, Modbus Tx and Modbus Tx Enable  _OR_   TCP Rx, TCP Tx  pins are defined. i.e. 
   - BL0939 (26,25,4) and (Modbus pins 1,3,5)   => works OK
   - BL0939 (26,25,4) and (Serial TCP Bridge 16,17)  => works OK
- if both remaining serial ports have pins assigned the sensor data is all zero for voltage, current and power
   - BL0939 (26,25,4) and (Modbus pins 1,3,5) and (Serial TCP Bridge 16,17)    =>  BL0939  doesn't work
   - there are no BL9 RX lines logged with two configured serial ports

### ADDITIONAL CONTEXT
Maybe triggering data for the BL0939 is not sent via the correct Tx wire (a guess).

**Comment by Jason2866** [comment]
Since BL0939 is not an typical DIY chip and needs AC mains by design, we won't add features to use more serial ports.

**Reply by thomas374b** [reply]
This is not a feature request, this is a bug report. 
- if serial ports are available, one may configure them, and if both are configured, BL0939 stops working

- please describe what is a "typical DIY chip" ? In my case I dismantled a SonoffDualR3 and equipped it with an ESP32 
  (BTW: I do _ANY_ chip when I try to recycle/repair old hardware)

- I'm already able to fix this issue by myself in my forked tasmota repo. I might keep this #issue/#discussion number for my upcoming pull request(s) ...  ;-)

**Reply by barbudor** [reply]
You should NOT connect any external sensor or equipment to a device with power measurment
Due to the design of power measurment circuits (whether BL09xx or any others) the ESP is at MAINS LEVEL

So whether this is a bug or not, just DON'T DO THAT
Keep your Sonoff closed and with no connection to external devices

**Reply by thomas374b** [reply]
Sorry, I reworked it already completly (BTW: with opto couplers, the device has all screws already tightened in its new housing for 6 weeks and is operating)

I know about MAINS LEVEL. Yes the ESP is at MAINS LEVEL if one follows the (chinese typical NPN instead of PNP) markings.

It is not on MAINS when you swap the wires. Backdraw, the load will be switched in the return path to ground.

**Reply by barbudor** [reply]
Wrong

**Comment by Jason2866** [comment]
To be 100% clear. We see this as security feature. So no PR will be merged to change this.

**Reply by thomas374b** [reply]
Maybe the problem is not only related to the BL09xx but to the handling of the serial ports in general.  I was just offering a code correction. If you say you don't need PRs, Thank you I will keep my time on doing other stuff.

Security is not maintained by keeping errors in software or by rejecting code changes

One can get electrocuted with or without correct or incorrect code.
